vcd/resource_vcd_vm_sizing_policy_test.go

```diff
@@ -0,0 +1,437 @@
+					resource.TestCheckOutput("limit_in_mb", params["MemoryLimit"].(string)),
+					resource.TestCheckOutput("memory_reservation_guarantee", params["MemoryReservation"].(string)),
+				),
+			},
+		},
+	})
+}
+
+func importStateVmSizingPolicyByIdOrName(testConfig TestConfig, resourceName string, byId bool) resource.ImportStateIdFunc {
+	return func(s *terraform.State) (string, error) {
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return "", fmt.Errorf("not found: %s", resourceName)
+		}
+		var identifier string
+		if byId {
+			identifier = rs.Primary.ID
+		} else {
+			identifier = rs.Primary.Attributes["name"]
+		}
+
+		if testConfig.VCD.Org == "" || identifier == "" {
+			return "", fmt.Errorf("missing information to generate import path %s, %s", testConfig.VCD.Org, identifier)
+		}
+		return testConfig.VCD.Org +
+			ImportSeparator +
+			identifier, nil
+	}
+}
+
+func testAccCheckVmSizingPolicyExists(name string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[name]
+		if !ok {
+			return fmt.Errorf("not found: %s", name)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("no VM sizing policy ID is set")
+		}
+
+		conn := testAccProvider.Meta().(*VCDClient)
+
+		adminOrg, err := conn.GetAdminOrg(testConfig.VCD.Org)
+		if err != nil {
+			return fmt.Errorf(errorRetrievingOrg, testConfig.VCD.Org+" and error: "+err.Error())
+		}
+
+		_, err = adminOrg.GetVdcComputePolicyById(rs.Primary.ID)
+		if err != nil {
+			return fmt.Errorf("VM sizing policy %s does not exist (%s)", rs.Primary.Attributes["name"], err)
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckVmSizingPolicyDestroyed(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*VCDClient)
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "vcd_org_vdc" && rs.Primary.Attributes["name"] != TestVmPolicy {
+			continue
+		}
+
+		adminOrg, err := conn.GetAdminOrg(testConfig.VCD.Org)
+		if err != nil {
+			return fmt.Errorf(errorRetrievingOrg, testConfig.VCD.Org+" and error: "+err.Error())
+		}
+
+		_, err = adminOrg.GetVdcComputePolicyById(rs.Primary.ID)
+
+		if err == nil {
+			return fmt.Errorf("VM sizing policy %s still exists", rs.Primary.ID)
+		}
+	}
+
+	return nil
+}
+
+func init() {
+	testingTags["vdc"] = "resource_vcd_org_vdc_test.go"
+}
+
+const testAccCheckVmSizingPolicy_basic = `
+resource "vcd_vm_sizing_policy" "{{.PolicyName}}_1" {
+  org         = "{{.OrgName}}"
+  name        = "{{.PolicyName}}_1"
+  description = "{{.Description}}_1"
+}
+
+resource "vcd_vm_sizing_policy" "{{.PolicyName}}_2" {
+  org         = "{{.OrgName}}"
+  name        = "{{.PolicyName}}_2"
+  description = "{{.Description}}_2"
+
+  cpu {
+    shares                = "{{.CpuShare}}"
+    limit_in_mhz          = "{{.CpuLimit}}"
+    count                 = "{{.CpuCount}}"
+    speed_in_mhz          = "{{.CpuSpeed}}"
+    cores_per_socket      = "{{.CoresPerSocket}}"
```

Comment:
I'm hitting an issue when running 2nd apply.

This is my resource:
```
resource "vcd_vm_sizing_policy" "min_size" {
  name        = "min-size"
  description = "smallest size"

  cpu {
    limit_in_mhz          = "2400"
    count                 = "1"
    speed_in_mhz          = "2500"
    cores_per_socket      = "1"
    reservation_guarantee = "0.55"
  }

  memory {
    size_in_mb            = "512"
    limit_in_mb           = "640"
    reservation_guarantee = "0.3"
  }
}
```

And this what happens after 2nd apply:

```
# vcd_vm_sizing_policy.min_size will be updated in-place
  ~ resource "vcd_vm_sizing_policy" "min_size" {
        description = "smallest size"
        id          = "urn:vcloud:vdcComputePolicy:8430ffb8-0649-4ed5-b69a-c946a7b4a53c"
        name        = "min-size"

      ~ cpu {
          + cores_per_socket      = "1"
          + count                 = "1"
            limit_in_mhz          = "2400"
            reservation_guarantee = "0.55"
            speed_in_mhz          = "2500"
        }

        memory {
            limit_in_mb           = "640"
            reservation_guarantee = "0.3"
            size_in_mb            = "512"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

1. Why?
2. Do we have double-apply testing enabled for this test?

CC: @dataclouder

Reply:
p.s. And this what happens if I say "yes" to the above apply:

```
Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vcd_vm_sizing_policy.min_size: Modifying... [id=urn:vcloud:vdcComputePolicy:8430ffb8-0649-4ed5-b69a-c946a7b4a53c]

Error: error updating VM sizing policy min-size, err: only name and description are updatable for VM sizing policy

  on radio2020.tf line 42, in resource "vcd_vm_sizing_policy" "min_size":
  42: resource "vcd_vm_sizing_policy" "min_size" {
```

Reply:
> p.s. And this what happens if I say "yes" to the above apply (in effect, this blocks me from using the feature):

Looks like it could use `ForceNew`

Reply:
I will add ForceNew, good idea. This will have one case with error if Policy will be already used - it will fail.

Reply:
> I'm hitting an issue when running 2nd apply.
> 
> This is my resource:
> 
> ```
> resource "vcd_vm_sizing_policy" "min_size" {
>   name        = "min-size"
>   description = "smallest size"
> 
>   cpu {
>     limit_in_mhz          = "2400"
>     count                 = "1"
>     speed_in_mhz          = "2500"
>     cores_per_socket      = "1"
>     reservation_guarantee = "0.55"
>   }
> 
>   memory {
>     size_in_mb            = "512"
>     limit_in_mb           = "640"
>     reservation_guarantee = "0.3"
>   }
> }
> ```
> 
> And this what happens after 2nd apply:
> 
> ```
> # vcd_vm_sizing_policy.min_size will be updated in-place
>   ~ resource "vcd_vm_sizing_policy" "min_size" {
>         description = "smallest size"
>         id          = "urn:vcloud:vdcComputePolicy:8430ffb8-0649-4ed5-b69a-c946a7b4a53c"
>         name        = "min-size"
> 
>       ~ cpu {
>           + cores_per_socket      = "1"
>           + count                 = "1"
>             limit_in_mhz          = "2400"
>             reservation_guarantee = "0.55"
>             speed_in_mhz          = "2500"
>         }
> 
>         memory {
>             limit_in_mb           = "640"
>             reservation_guarantee = "0.3"
>             size_in_mb            = "512"
>         }
>     }
> 
> Plan: 0 to add, 1 to change, 0 to destroy.
> ```
> 
>     1. Why?
> 
>     2. Do we have double-apply testing enabled for this test?
> 
> 
> CC: @dataclouder

This is a very strange problem. I have replicated it with exactly such config. However I have put in place a very similar config (with same values for `count` and `cores_per_socket`. ) it does not ask for changes:
```
resource "vcd_vm_sizing_policy" "TestAccVcdVmSizingPolicy_4" {
  org         = "ss"
  name        = "TestAccVcdVmSizingPolicy_4"
  description = "TestAccVcdVmSizingPolicyDescription_4"

  cpu {
    count                 = "1"
    cores_per_socket      = "1"
    shares                = "886"
    limit_in_mhz          = "2400"
    speed_in_mhz          = "2500"
    reservation_guarantee = "0.55"
  }

  memory {
    shares                = "1580"
    size_in_mb            = "3200"
    limit_in_mb           = "2800"
    reservation_guarantee = "0.3"
  }
}
```

Reply:
ForceNew added. Bug fixed.